src/main/java/com/wiley/autotest/selenium/elements/upgrade/BaseTeasyElement.java

```diff
@@ -0,0 +1,495 @@
+    public List<TeasyElement> elements(By by, SearchStrategy strategy) {
+        return customContextFinder(strategy).visibleElements(by);
+    }
+
+    @Override
+    public TeasyElement domElement(By by) {
+        return contextFinder.presentInDomElement(by);
+    }
+
+    @Override
+    public TeasyElement domElement(By by, SearchStrategy strategy) {
+        return customContextFinder(strategy).presentInDomElement(by);
+    }
+
+    @Override
+    public List<TeasyElement> domElements(By by) {
+        return contextFinder.presentInDomElements(by);
+    }
+
+    @Override
+    public List<TeasyElement> domElements(By by, SearchStrategy strategy) {
+        return customContextFinder(strategy).presentInDomElements(by);
+    }
+
+    TeasyElementFinder customContextFinder(SearchStrategy strategy) {
+        return new TeasyElementFinder(getWebDriver(), strategy, this);
+    }
+
+    @Override
+    public void click() {
+        try {
+            try {
+                wrappedElement.click();
+            } catch (StaleElementReferenceException e) {
+                clickForStaleElement();
+            } catch (UnhandledAlertException ignored) {
+                new Report("*****UnhandledAlertException***** during click! Doing nothing just trying to continue the test. Locator=" + locator
+                        .getLocator()).jenkins();
+            } catch (UnreachableBrowserException ignored) {
+                //doing this because for FireFox for some reason browser becomes unresponsive after click
+                //but actually it is alive so it worth to try to continue test
+                //it will fail on the next method after click if some real error happened
+                new Report("*****ERROR*****UnreachableBrowserException***** during click! Doing nothing just trying to continue the test. Locator=" + locator
+                        .getLocator()).jenkins();
+            } catch (ElementNotVisibleException needToScroll) {
+                clickForNeedToScroll();
+            } catch (WebDriverException ignoredOrNeedToScroll) {
+                clickForIgnoredScroll(ignoredOrNeedToScroll);
+            }
+        } catch (Exception e) {
+            new Report("*****UNKNOWN ERROR*****Exception***** DURING CLICK LOGIC. SHOULD BE REFACTORED!!!! Locator=" + locator.getLocator(), e).jenkins();
+        }
+    }
+
+    private void clickForStaleElement() {
+        againLocate();
+        click();
+    }
+
+    private void clickForNeedToScroll() {
+        new Report("ElementNotVisibleException***** during click! Scrolling to element and trying again ---Locator=" + locator.getLocator()).jenkins();
+        increment();
+        scrollIntoView(wrappedElement);
+        scrollToElementLocation(wrappedElement);
+        click();
+    }
+
+    private void clickForIgnoredScroll(WebDriverException ignoredOrNeedToScroll) {
+        new Report("*****WebDriverException***** during click!-----Locator=" + locator.getLocator()).jenkins();
```

Comment:
Very strange statement. Create a instance of class to call single method. Maybe we need a static method here 
Report.jenkins("message text") to help garbage collector

Reply:
@dpavlov it was my attempt to code in a more OOP way... where statics does not fit:). I got your point, let's discuss to make sure that I need to change this ;)